src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryPartitionedUserTokenCacheAccessor.cs

```diff
@@ -0,0 +1,259 @@
+            new ConcurrentDictionary<string, ConcurrentDictionary<string, MsalAccountCacheItem>>(1, 1);
+
+        internal /* internal for test only */  readonly ConcurrentDictionary<string, MsalAppMetadataCacheItem> AppMetadataDictionary =
+           new ConcurrentDictionary<string, MsalAppMetadataCacheItem>(1, 1);
+
+        protected readonly ICoreLogger _logger;
+
+        public InMemoryPartitionedUserTokenCacheAccessor(ICoreLogger logger)
+        {
+            _logger = logger ?? throw new System.ArgumentNullException(nameof(logger));
+        }
+
+        #region Add
+        public void SaveAccessToken(MsalAccessTokenCacheItem item)
+        {
+            string itemKey = item.GetKey().ToString();
+
+            string partitionKey = CacheKeyFactory.GetKeyFromCachedItem(item);
+
+            AccessTokenCacheDictionary
+                .GetOrAdd(partitionKey, new ConcurrentDictionary<string, MsalAccessTokenCacheItem>())[itemKey] = item; // if a conflict occurs, pick the latest value
+        }
+
+        public void SaveRefreshToken(MsalRefreshTokenCacheItem item)
+        {
+            string itemKey = item.GetKey().ToString();
+            string partitionKey = CacheKeyFactory.GetKeyFromCachedItem(item);
+            RefreshTokenCacheDictionary
+                .GetOrAdd(partitionKey, new ConcurrentDictionary<string, MsalRefreshTokenCacheItem>())[itemKey] = item;
+        }
+
+        public void SaveIdToken(MsalIdTokenCacheItem item)
+        {
+            string itemKey = item.GetKey().ToString();
+            string partitionKey = CacheKeyFactory.GetKeyFromCachedItem(item);
+
+            IdTokenCacheDictionary
+                .GetOrAdd(partitionKey, new ConcurrentDictionary<string, MsalIdTokenCacheItem>())[itemKey] = item;
+        }
+
+        public void SaveAccount(MsalAccountCacheItem item)
+        {
+            string itemKey = item.GetKey().ToString();
+            string partitionKey = CacheKeyFactory.GetKeyFromCachedItem(item);
+
+            AccountCacheDictionary
+                .GetOrAdd(partitionKey, new ConcurrentDictionary<string, MsalAccountCacheItem>())[itemKey] = item;
+        }
+
+        public void SaveAppMetadata(MsalAppMetadataCacheItem item)
+        {
+            string key = item.GetKey().ToString();
+            AppMetadataDictionary[key] = item;
+        }
+        #endregion
+
+        #region Get
+
+        public MsalIdTokenCacheItem GetIdToken(MsalAccessTokenCacheItem accessTokenCacheItem)
+        {
+            string partitionKey = CacheKeyFactory.GetIdTokenKeyFromCachedItem(accessTokenCacheItem);
+
+            IdTokenCacheDictionary.TryGetValue(partitionKey, out var partition);
+            if (partition != null && partition.TryGetValue(accessTokenCacheItem.GetIdTokenItemKey().ToString(), out var idToken))
+            {
+                return idToken;
+            }
+
+            _logger.WarningPii(
```

Comment:
the messages are different, and also not sure why they are pii

Reply:
GetKey contains some light PII, such as the home account id